File: src/main/scala/abc133/MainE.kt
package abc133

import java.util.*


data class Edge(val a: Int, val b: Int)

data class GF(val x: Long) {
    companion object {
        const val MOD: Long = (1e9 + 7).toLong()

        fun powGF(x: GF, t: Long): GF {
            if (t == 0L) return GF(1)
            val y = powGF(x, t / 2L)
            return if (t % 2L == 0L) {
                y * y
            } else {
                x * y * y
            }
        }
    }

    operator fun plus(that: GF): GF {
        return GF((this.x + that.x) % MOD)
    }

    operator fun minus(that: GF): GF {
        return GF((this.x - that.x + MOD) % MOD)
    }

    operator fun times(that: GF): GF {
        return GF((this.x * that.x) % MOD)
    }

    operator fun unaryMinus(): GF {
        return GF((-this.x + MOD) % MOD)
    }

    fun inv(): GF {
        return powGF(this, MOD - 2)
    }

    operator fun div(that: GF): GF {
        return this * that.inv()
    }
}

fun main(args: Array<String>) {
    val (n, k) = readLine()!!.split(" ").map(String::toInt)
    val edges = Array(n - 1) {
        val (a, b) = readLine()!!.split(" ").map(String::toInt)
        Edge(a - 1, b - 1)
    }

    val nextList = Array(n) {
        mutableListOf<Int>()
    }
    for (edge in edges) {
        nextList[edge.a].add(edge.b)
        nextList[edge.b].add(edge.a)
    }

    val queue: Queue<Int> = ArrayDeque()
    val visited = Array(n) { false }

    val numNextVisited = Array(n) { 0 }
    val score = Array(n) { -123 }

    queue.add(0)
    visited[0] = true
    score[0] = k

    while (queue.isNotEmpty()) {
        val current = queue.remove()
        for (next in nextList[current]) {
            if (!visited[next]) {
                queue.add(next)
                visited[next] = true

                score[next] = k - (1 + numNextVisited[current])
                numNextVisited[next]++
                numNextVisited[current]++
            }
        }
    }
    val result = score.map { GF(it.toLong()) }.fold(GF(1)) { x, y -> x * y }.x
    println(result)

}

File: src/main/scala/abc133/MainE.scala
package abc133

import java.util
import java.util.Scanner

import scala.collection.mutable.ArrayBuffer


object MainE {

  case class Edge(from: Int, to: Int)

  case class RR(representative: Long) extends AnyVal {
    def +(that: RR): RR = RR((this.representative + that.representative) % RR.mod)

    def -(that: RR): RR = RR((this.representative - that.representative + RR.mod) % RR.mod)

    def *(that: RR): RR = RR((this.representative * that.representative) % RR.mod)

    def unary_- : RR = RR((-this.representative + RR.mod) % RR.mod)
  }


  object RR {
    val mod: Long = (1e9+7).toLong

    val _0: RR = RR(0)
    val _1: RR = RR(1)
  }


  def read() = {
    val sc = new Scanner(System.in)
    val n, k = sc.nextInt()
    val edges = IndexedSeq.fill(n - 1) {
      val from = sc.nextInt() - 1
      val to = sc.nextInt() - 1
      Edge(from, to)
    }
    (n, k, edges)
  }

  def solve(n: Int, k: Long, edges: IndexedSeq[Edge]): Long = {
    val nextList = IndexedSeq.fill(n)(ArrayBuffer.empty[Int])
    for (edge <- edges) {
      nextList(edge.from).append(edge.to)
      nextList(edge.to).append(edge.from)
    }
    val numNextVisited = Array.fill(n)(0)
    val visited = Array.fill(n)(false)
    val queue: util.Queue[Int] = new util.ArrayDeque[Int]()
    val score = Array.fill(n)(-123L)


    queue.add(0)
    score(0) = k
    visited(0) = true

    while (!queue.isEmpty) {
      val current = queue.remove()
      for (next <- nextList(current)) {
        if (!visited(next)) {
          score(next) = k- (1 + numNextVisited(current))
          queue.add(next)
          visited(next) = true

          numNextVisited(next) += 1
          numNextVisited(current) += 1
        }
      }
    }
    score.map(RR(_)).foldLeft(RR(1))(_*_).representative
  }

  def main(args: Array[String]): Unit = {
    val (n, k, edges) = read()
    println(solve(n,k,edges))
  }
}